core/fullHistory/historyRepository.go

```diff
@@ -101,105 +95,220 @@ func (hp *historyProcessor) PutTransactionsData(historyTxsData *HistoryTransacti
+func (hp *historyProcessor) GetEpochByHash(hash []byte) (uint32, error) {
+	return hp.epochByHashIndex.getEpochByHash(hash)
+}
+
+func (hp *historyProcessor) RegisterToBlockTracker(blockTracker BlockTracker) {
+	if check.IfNil(blockTracker) {
+		log.Error("RegisterToBlockTracker(): blockTracker is nil")
+		return
+	}
+
+	blockTracker.RegisterCrossNotarizedHeadersHandler(hp.onNotarizedBlocks)
+	blockTracker.RegisterSelfNotarizedHeadersHandler(hp.onNotarizedBlocks)
+}
+
+func (hp *historyProcessor) onNotarizedBlocks(shardID uint32, headers []data.HeaderHandler, headersHashes [][]byte) {
+	log.Trace("onNotarizedBlocks()", "shardID", shardID, "len(headers)", len(headers))
+
+	if shardID != core.MetachainShardId {
+		return
+	}
+
+	for i := 0; i < len(headers); i++ {
+		header := headers[i]
+		headerHash := headersHashes[i]
+
+		metaBlock, ok := header.(*block.MetaBlock)
+		if !ok {
+			// Question for review: why, on every round, there's a header of type = *block.Header (even if shardID = metachain)?
+			log.Debug("onNotarizedBlocks(): cannot convert to *block.Metablock", "type", fmt.Sprintf("%T", header))
+			return
+		}
+
+		log.Trace("onNotarizedBlocks()", "index", i, "len(ShardInfo)", len(metaBlock.ShardInfo))
+		for _, shardData := range metaBlock.ShardInfo {
+			hp.onNotarizedBlock(metaBlock.GetNonce(), headerHash, shardData)
+		}
+	}
+}
+
+func (hp *historyProcessor) onNotarizedBlock(metaBlockNonce uint64, metaBlockHash []byte, blockHeader block.ShardData) {
+	log.Trace("onNotarizedBlock()",
+		"metaBlockNonce", metaBlockNonce,
+		"nonce", blockHeader.Nonce,
+		"shard", blockHeader.ShardID,
+		"len(miniblocks)", len(blockHeader.ShardMiniBlockHeaders),
+	)
+
+	for _, miniblockHeader := range blockHeader.GetShardMiniBlockHeaders() {
+		hp.onNotarizedMiniblock(metaBlockNonce, metaBlockHash, blockHeader, miniblockHeader)
+	}
 }
 
-// GetEpochForHash will return epoch for a given hash
-func (hp *historyProcessor) GetEpochForHash(hash []byte) (uint32, error) {
-	return hp.hashEpochStorer.GetEpoch(hash)
+func (hp *historyProcessor) onNotarizedMiniblock(metaBlockNonce uint64, metaBlockHash []byte, blockHeader block.ShardData, miniblockHeader block.MiniBlockHeader) {
+	miniblockHash := miniblockHeader.Hash
+	isIntra := miniblockHeader.SenderShardID == miniblockHeader.ReceiverShardID
+	notarizedAtSource := miniblockHeader.SenderShardID == blockHeader.ShardID
+
+	iDontCare := miniblockHeader.SenderShardID != hp.selfShardID && miniblockHeader.ReceiverShardID != hp.selfShardID
+	if iDontCare {
+		log.Trace("onNotarizedMiniblock(): skipping",
+			"miniblockHash", miniblockHash,
+			"sender shard", miniblockHeader.SenderShardID,
+			"receiver shard", miniblockHeader.ReceiverShardID,
+		)
+		return
+	}
+
+	metadata, err := hp.getMiniblockMetadataByMiniblockHash(miniblockHash)
+	if err != nil {
+		// Question for review: is there a way to be notified about the "source notarization", but at destination (cross-shard miniblock)?
```

Comment:
R: when metachain block notarizes the miniblock with self as sender, but for the destination shard (destination shard header).

Reply:
Fixed - we hold on to notifications of type "notarized at source" and use them at "commit at destination" time.